<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [44]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [45]:
import folium
import pandas as pd

In [46]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


In [7]:
## Task 1: Mark all launch sites on a map

First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [47]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [48]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [49]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [50]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#2A02D2', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)
site_map.save('nasa_jsc_map1.html')
site_map

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [51]:
import folium

# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# Create a map centered around the first launch site
m = folium.Map(location=[launch_sites_df["Lat"][0], launch_sites_df["Long"][0]], zoom_start=5)

# Create a circle and marker for each launch site
for index, row in launch_sites_df.iterrows():
    coordinate = [row["Lat"], row["Long"]]
    
    # Create a circle
    folium.Circle(
        location=coordinate,
        radius=1000,  # Radius in meters
        color='red',
        fill=True
    ).add_child(folium.Popup(row["Launch Site"])).add_to(m)
    
    # Create a marker
    folium.Marker(
        location=coordinate,
        icon=folium.DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html='<div style="font-size: 12px; color:#d35400;"><b>{}</b></div>'.format(row["Launch Site"])
        )
    ).add_to(m)

# Display the map
m.save('Launch_Site.html')
m

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


In [ ]:
# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [52]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [53]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
# Function to determine marker color based on class value
def get_marker_color(class_value):
    if class_value == 1:
        return 'green'
    else:
        return 'red'

# Apply the function to create the 'marker_color' column
spacex_df['marker_color'] = spacex_df['class'].apply(get_marker_color)

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [10]:
spacex_df.head()

Launch Site        Lat       Long  class marker_color
0  CCAFS LC-40  28.562302 -80.577356      0          red
1  CCAFS LC-40  28.562302 -80.577356      0          red
2  CCAFS LC-40  28.562302 -80.577356      0          red
3  CCAFS LC-40  28.562302 -80.577356      0          red
4  CCAFS LC-40  28.562302 -80.577356      0          red

In [54]:
# Add marker_cluster to current site_map
marker_cluster = MarkerCluster().add_to(site_map)

# For each row in spacex_df data frame
# Create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successful or failed
for index, record in spacex_df.iterrows():
    # Debug print statement
    # print(f"Adding marker for launch site: {record['Launch Site']} at coordinates: ({record['Lat']}, {record['Long']})")
    
    # Create a marker with the popup
    popup_text = f"Launch Site {record['Launch Site']}"
    marker = folium.Marker(
        location=[record["Lat"], record["Long"]],
        icon=folium.Icon(color='white', icon_color=record['marker_color']),
        popup=popup_text
    )
    marker_cluster.add_child(marker)

# Display the map
site_map.save('current_site.html')
site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


In [ ]:
# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [55]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map.save('mouse_position.html')
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [56]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [ ]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

In [ ]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

In [35]:
# Debug print statements
print(f"Launch site coordinates: ({launch_site_lat}, {launch_site_lon})")
print(f"Coastline coordinates: ({coastline_lat}, {coastline_lon})")
print(f"Calculated distance: {distance_coastline} KM")

Launch site coordinates: (28.56230197, -80.57735648)
Coastline coordinates: (28.56367, -80.57163)
Calculated distance: 0.5797581813109574 KM


In [70]:
import folium
from folium.features import DivIcon
from math import sin, cos, sqrt, atan2, radians

# Define the calculate_distance function
def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

# Pick the launch site from the DataFrame
selected_site = 'CCAFS SLC-40'
launch_row = launch_sites_df[launch_sites_df['Launch Site'] == selected_site].iloc[0]

# Launch Site Coordinates
launch_site_lat = launch_row['Lat']
launch_site_lon = launch_row['Long']

coastline_lat = 28.56367 
coastline_lon = -80.5680

# Calculate the distance
distance = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

# Debug print statement
print(f"Calculated distance: {distance} KM")

# Create the map centered at the launch site
site_map = folium.Map(location=[launch_site_lat, launch_site_lon], zoom_start=14)

# Add Launch Site Marker
folium.Marker(
    location=[launch_site_lat, launch_site_lon],
    popup=f"Launch Site: {selected_site}",
    icon=folium.Icon(color='blue')
).add_to(site_map)

# Add Coastline Marker
folium.Marker(
    location=[coastline_lat, coastline_lon],
    popup="Closest Coastline Point",
    icon=folium.Icon(color='green')
).add_to(site_map)

# Draw Polyline
coordinates = [[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]]
lines = folium.PolyLine(locations=coordinates, weight=1, color='blue')
site_map.add_child(lines)

# Add Distance Label at Midpoint
if distance is not None:
    distance_marker = folium.Marker(
        location=[(launch_site_lat + coastline_lat) / 2, (launch_site_lon + coastline_lon) / 2],
        icon=DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html='<div style="font-size: 12px; color:#d35400;"><b>{:10.2f} KM</b></div>'.format(distance)
        )
    )
    site_map.add_child(distance_marker)
else:
    print("Error: Distance is None. Please check the calculate_distance function.")

# Save or display the map
site_map.save('dynamic_launchsite_to_coastline.html')
site_map

Calculated distance: 0.8632486563613652 KM


*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [61]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [ ]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site


In [72]:
import folium
from folium.features import DivIcon
from math import sin, cos, sqrt, atan2, radians

# Define the calculate_distance function
def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

# Example coordinates for illustration
city_lat, city_lon = 28.5384, -81.38900          # closest city: Orlando
railway_lat, railway_lon = 28.55129, -80.80045   # closest railway point: Florida Eastcoast railway
highway_lat, highway_lon = 28.52542, -80.62929   # closest highway point: FL405 NASA Parkway

# # From West coast
# launch_site_lat_w, launch_site_lon_w = 34.6320, -120.6107 # VAFB SLC4E
# city_lat_w, city_lon_w = 34.6392, -120.4579 # Lompoc, California 

# Calculate distances from launch site
distance_city = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)
distance_city_w = calculate_distance(launch_site_lat_w, launch_site_lon_w, city_lat_w, city_lon_w)
distance_railway = calculate_distance(launch_site_lat, launch_site_lon, railway_lat, railway_lon)
distance_highway = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)

# # Pick the launch site from the DataFrame
# selected_site_w = 'VAFB SLC-4E'
# launch_row = launch_sites_df[launch_sites_df['Launch Site'] == selected_site_w].iloc[0]
# # Create the map centered at the launch site
# site_map = folium.Map(location=[launch_site_lat, launch_site_lon], zoom_start=14)

# # Add Launch Site Marker
# folium.Marker(
#     location=[launch_site_lat, launch_site_lon],
#     popup=f"Launch Site: {selected_site_w}",
#     icon=folium.Icon(color='red')
# ).add_to(site_map)

# Add City Marker East Coast
folium.Marker(
    location=[city_lat, city_lon],
    popup="Closest City East Coast",
    icon=folium.Icon(color='purple')
).add_to(site_map)

# # Add City Marker West Coast
# folium.Marker(
#     location=[city_lat_w, city_lon_w],
#     popup="Closest City West Coast",
#     icon=folium.Icon(color='gray')
# ).add_to(site_map)

# Add Railway Marker
folium.Marker(
    location=[railway_lat, railway_lon],
    popup="Closest Railway",
    icon=folium.Icon(color='green')
).add_to(site_map)

# Add Highway Marker
folium.Marker(
    location=[highway_lat, highway_lon],
    popup="Closest Highway",
    icon=folium.Icon(color='blue')
).add_to(site_map)

# Draw PolyLines to each proximity point
folium.PolyLine(locations=[[launch_site_lat, launch_site_lon], [city_lat, city_lon]], color='purple', weight=1).add_to(site_map)
# folium.PolyLine(locations=[[launch_site_lat_w, launch_site_lon_w], [city_lat_w, city_lon_w]], color='gray', weight=1).add_to(site_map)
folium.PolyLine(locations=[[launch_site_lat, launch_site_lon], [railway_lat, railway_lon]], color='green', weight=1).add_to(site_map)
folium.PolyLine(locations=[[launch_site_lat, launch_site_lon], [highway_lat, highway_lon]], color='blue', weight=1).add_to(site_map)

# Add Distance Markers for each proximity
folium.Marker(
    location=[(launch_site_lat + city_lat) / 2, (launch_site_lon + city_lon) / 2],
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:purple;"><b>{:10.2f} KM</b></div>'.format(distance_city),
    )
).add_to(site_map)

# folium.Marker(
#     location=[(launch_site_lat_w + city_lat_w) / 2, (launch_site_lon_w + city_lon_w) / 2],
#     icon=DivIcon(
#         icon_size=(20,20),
#         icon_anchor=(0,0),
#         html='<div style="font-size: 12px; color:gray;"><b>{:10.2f} KM</b></div>'.format(distance_city_w),
#     )
# ).add_to(site_map)

folium.Marker(
    location=[(launch_site_lat + railway_lat) / 2, (launch_site_lon + railway_lon) / 2],
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:green;"><b>{:10.2f} KM</b></div>'.format(distance_railway),
    )
).add_to(site_map)

folium.Marker(
    location=[(launch_site_lat + highway_lat) / 2, (launch_site_lon + highway_lon) / 2],
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:blue;"><b>{:10.2f} KM</b></div>'.format(distance_highway),
    )
).add_to(site_map)

# Save the map
site_map.save('launchsite_with_proximities.html')
site_map

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


Are launch sites in close proximity to railways? 21.89 kms from launch site CCAFS SLC-40 to Florida Eastcoast Railway

Are launch sites in close proximity to highways? 6.63 kms from launch site CCAFS SLC-40 to Highway - FL405 Nasa Parkway

Are launch sites in close proximity to coastline? 0.86 kms from launch site CCAFS

Do launch sites keep certain distance away from cities? 79.40 kms from launch site CCAFS SLC-40 to Orlando City

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
